src/autotype/AutoType.cpp

```diff
@@ -399,437 +467,224 @@ void AutoType::restoreWindowState()
+                if (repeat > maxRepetition && error) {
+                    *error = tr("Too many repetitions detected, max is %1: %2").arg(maxRepetition).arg(fullPlaceholder);
+                    return {};
+                }
+                auto action = QSharedPointer<AutoTypeKey>::create(g_placeholderToKey[placeholder], modifiers);
+                for (int i = 1; i <= repeat && i <= maxRepetition; ++i) {
+                    actions << action;
+                }
+            } else if (placeholder.startsWith("delay=")) {
+                // Change keypress delay
+                int delay = placeholder.replace("delay=", "").toInt();
+                if (delay > maxTypeDelay && error) {
+                    *error = tr("Very slow key press detected, max is %1: %2").arg(maxTypeDelay).arg(fullPlaceholder);
+                    return {};
+                }
+                actions << QSharedPointer<AutoTypeDelay>::create(qBound(0, delay, maxTypeDelay), true);
+            } else if (placeholder == "delay") {
+                // Mid typing delay (wait)
+                if (repeat > maxWaitDelay && error) {
+                    *error = tr("Very long delay detected, max is %1: %2").arg(maxWaitDelay).arg(fullPlaceholder);
+                    return {};
+                }
+                actions << QSharedPointer<AutoTypeDelay>::create(qBound(0, repeat, maxWaitDelay));
+            } else if (placeholder == "clearfield") {
+                // Platform-specific field clearing
+                actions << QSharedPointer<AutoTypeClearField>::create();
+            } else if (placeholder == "totp") {
+                // Entry totp (requires special handling)
+                QString totp = entry->totp();
+                for (const auto& ch : totp) {
+                    actions << QSharedPointer<AutoTypeKey>::create(ch);
+                }
+            } else if (placeholder == "pickchars") {
+                if (error) {
+                    // Ignore this if we are syntax checking
+                    continue;
+                }
+                // Show pickchars dialog for entry's password
+                auto password = entry->resolvePlaceholder(entry->password());
+                if (!password.isEmpty()) {
+                    PickcharsDialog pickcharsDialog(password);
+                    if (pickcharsDialog.exec() == QDialog::Accepted && !pickcharsDialog.selectedChars().isEmpty()) {
+                        auto chars = pickcharsDialog.selectedChars();
+                        auto iter = chars.begin();
+                        while (iter != chars.end()) {
+                            actions << QSharedPointer<AutoTypeKey>::create(*iter);
+                            ++iter;
+                            if (pickcharsDialog.pressTab() && iter != chars.end()) {
+                                actions << QSharedPointer<AutoTypeKey>::create(g_placeholderToKey["tab"]);
+                            }
```

Comment:
No because I use the fact that we are not at the end to add a tab between key presses. With a range loop I have to add logic after the loop to back out a tab press.